lib/internal/per_context/primordials.js

```diff
@@ -384,5 +386,23 @@ primordials.SafeWeakRef = makeSafe(
   }
 );
 
+const SafePromise = makeSafe(
+  Promise,
+  class SafePromise extends Promise {
+    // eslint-disable-next-line no-useless-constructor
+    constructor(executor) { super(executor); }
+  }
+);
+
+primordials.PromisePrototypeCatch = (thisPromise, onRejected) =>
+  PromisePrototypeThen(thisPromise, undefined, onRejected);
+
+primordials.PromisePrototypeFinally = (thisPromise, onFinally) =>
+  new Promise((a, b) =>
+    new SafePromise((a, b) => PromisePrototypeThen(thisPromise, a, b))
+      .finally(onFinally)
+      .then(a, b)
```

Comment:
I think this creates too many promises to be usable in any hot code paths. I would recommend not adding this.

Reply:
Aren't promises antithetical to hot code path since they are asynchronous? For info, `PromisePrototypeFinally` is only used in `fs/promises` and `run_main` currently. Would you prefer if I added a lint rule to discourage its use so it doesn't end up in a hot code path?

Reply:
I'm not sure if this allocates 3 or 4 promises instead of 1.
Using this specific code will only create problems.

Reply:
I'd also be more in favor of discouraging the use of the `catch` and `finally` methods in core:
- catch can be easily replaced with `then(undefined, fn)`
- finally may be replaceable with async functions

Reply:
I've removed used of `PromisePrototypeFinally` where it was possible, and rename the function to `SafePromisePrototypeFinally` to highlight the fact it is not a simple bridge to the actual primordial method. I plan to add it to the list of "problematic" primordials to avoid in hot code path in https://github.com/nodejs/node/pull/38635.